staging/vhost-device-sound/src/stream.rs

```diff
@@ -236,49 +231,98 @@ impl Default for PcmParams {
 }
 
 pub struct Buffer {
-    // TODO: to make private and add len usize
-    pub data_descriptor: virtio_queue::Descriptor,
+    pub bytes: Vec<u8>,
     pub pos: usize,
+    data_descriptor: virtio_queue::Descriptor,
     pub message: Arc<IOMessage>,
+    populated: bool,
+    direction: Direction,
 }
 
 impl std::fmt::Debug for Buffer {
     fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
         fmt.debug_struct(stringify!(Buffer))
+            .field("bytes", &self.bytes.len())
             .field("pos", &self.pos)
+            .field("populated", &self.populated)
+            .field("direction", &self.direction)
             .field("message", &Arc::as_ptr(&self.message))
             .finish()
     }
 }
 
 impl Buffer {
-    pub fn new(data_descriptor: virtio_queue::Descriptor, message: Arc<IOMessage>) -> Self {
+    pub fn new(
+        data_descriptor: virtio_queue::Descriptor,
+        message: Arc<IOMessage>,
+        direction: Direction,
+    ) -> Self {
         Self {
+            bytes: vec![],
             pos: 0,
             data_descriptor,
+            populated: false,
             message,
+            direction,
         }
     }
 
-    pub fn consume(&self, buf: &mut [u8]) -> Result<u32> {
-        let addr = self.data_descriptor.addr();
-        let offset = self.pos as u64;
-        let len = self
+    pub fn prepare_output(&mut self) -> Result<()> {
+        if self.populated {
+            return Ok(());
+        }
+        self.bytes = vec![0; self.data_descriptor.len() as usize];
+        let bytes_read = self
             .message
             .desc_chain
             .memory()
-            .read(
-                buf,
-                addr.checked_add(offset)
-                    .expect("invalid guest memory address"),
-            )
-            .map_err(|_| Error::DescriptorReadFailed)?;
-        Ok(len as u32)
+            .read(&mut self.bytes, self.data_descriptor.addr())
+            .map_err(|_| crate::Error::DescriptorReadFailed)?;
+        self.bytes.truncate(bytes_read);
+        self.populated = true;
+
+        Ok(())
+    }
+
+    pub fn prepare_input(&mut self) {
+        if self.populated {
+            return;
+        }
+        self.bytes = vec![0; self.data_descriptor.len() as usize];
```

Comment:
The `prepare_input()` function could also write directly to guest memory, so we wouldn't have to allocate bytes[]. It might be a different PR, though.

Reply:
@MatiasVara also maybe it would be possible to implement std::io::Write in an elegant way and do vectored read/writes from the descriptors.